File: engine/notifications.py
"""Notification CRUD helpers — used by api/routes/notifications.py and event hooks."""
import uuid
from contextlib import contextmanager
from db.conversation_store import _conn


def get_supervisor_ids() -> list[str]:
    """Return all active supervisor/admin user IDs for fan-out notifications."""
    with _conn() as conn:
        cur = conn.cursor()
        cur.execute(
            "SELECT id FROM users WHERE role IN ('supervisor', 'admin', 'super_admin') AND active = TRUE"
        )
        return [str(r["id"]) for r in cur.fetchall()]


def create_notification(
    user_id: str,
    role: str,
    type: str,
    priority: str,
    title: str,
    body: str,
    ticket_id: str | None = None,
) -> dict:
    """Insert a notification row and return it."""
    nid = str(uuid.uuid4())
    with _conn() as conn:
        cur = conn.cursor()
        cur.execute(
            """
            INSERT INTO notifications (id, user_id, role, type, priority, title, body, ticket_id)
            VALUES (%s, %s, %s, %s, %s, %s, %s, %s)
            RETURNING id, user_id, role, type, priority, title, body, ticket_id, read, created_at
            """,
            (nid, user_id, role, type, priority, title, body, ticket_id),
        )
        row = dict(cur.fetchone())
        row["created_at"] = row["created_at"].isoformat()
        return row


def fan_out_to_supervisors(
    type: str,
    priority: str,
    title: str,
    body: str,
    ticket_id: str | None = None,
) -> list[dict]:
    """Create a notification for every active supervisor/admin. Returns list of created rows."""
    supervisor_ids = get_supervisor_ids()
    results = []
    for uid in supervisor_ids:
        notif = create_notification(
            user_id=uid,
            role="supervisor",
            type=type,
            priority=priority,
            title=title,
            body=body,
            ticket_id=ticket_id,
        )
        results.append(notif)
    return results


def get_notifications(user_id: str, limit: int = 50) -> list[dict]:
    """Fetch the most recent notifications for a user."""
    with _conn() as conn:
        cur = conn.cursor()
        cur.execute(
            """
            SELECT id, user_id, role, type, priority, title, body, ticket_id, read, created_at
            FROM notifications
            WHERE user_id = %s
            ORDER BY created_at DESC
            LIMIT %s
            """,
            (user_id, limit),
        )
        rows = []
        for r in cur.fetchall():
            row = dict(r)
            row["created_at"] = row["created_at"].isoformat()
            rows.append(row)
        return rows


def mark_read(notif_id: str, user_id: str) -> bool:
    """Mark a single notification as read. Returns True if found and updated."""
    with _conn() as conn:
        cur = conn.cursor()
        cur.execute(
            "UPDATE notifications SET read = TRUE WHERE id = %s AND user_id = %s",
            (notif_id, user_id),
        )
        return cur.rowcount > 0


def mark_all_read(user_id: str) -> int:
    """Mark all unread notifications as read for a user. Returns count updated."""
    with _conn() as conn:
        cur = conn.cursor()
        cur.execute(
            "UPDATE notifications SET read = TRUE WHERE user_id = %s AND read = FALSE",
            (user_id,),
        )
        return cur.rowcount


def notification_recently_exists(ticket_id: str, notif_type: str, within_seconds: int = 3600) -> bool:
    """Return True if a notification of the given type for this ticket was created recently."""
    with _conn() as conn:
        cur = conn.cursor()
        cur.execute(
            """
            SELECT 1 FROM notifications
            WHERE ticket_id = %s AND type = %s
              AND created_at > NOW() - (%s * INTERVAL '1 second')
            LIMIT 1
            """,
            (ticket_id, notif_type, within_seconds),
        )
        return cur.fetchone() is not None

File: dashboard/server/src/routes/notifications.js
// /api/notifications
const router = require('express').Router();
const pool   = require('../db/pg');
const { authenticate } = require('../middleware/auth');
const { emitToAgent, emitToSupervisors } = require('../lib/sockets');
const { v4: uuidv4 } = require('uuid');

// ── GET /api/notifications ────────────────────────────────────────────────────
router.get('/', authenticate, async (req, res) => {
  try {
    const { rows } = await pool.query(
      `SELECT id, user_id, role, type, priority, title, body, ticket_id, read, created_at
       FROM notifications
       WHERE user_id = $1
       ORDER BY created_at DESC
       LIMIT 200`,
      [req.user.id]
    );
    res.json(rows);
  } catch (err) {
    console.error('[notifications] GET error:', err.message);
    res.status(500).json({ error: 'Server error' });
  }
});

// ── PATCH /api/notifications/:id/read ────────────────────────────────────────
router.patch('/:id/read', authenticate, async (req, res) => {
  try {
    const { rowCount } = await pool.query(
      `UPDATE notifications SET read = TRUE WHERE id = $1 AND user_id = $2`,
      [req.params.id, req.user.id]
    );
    if (rowCount === 0) return res.status(404).json({ error: 'Not found' });
    res.json({ status: 'read' });
  } catch (err) {
    console.error('[notifications] PATCH read error:', err.message);
    res.status(500).json({ error: 'Server error' });
  }
});

// ── PATCH /api/notifications/:id/unread ──────────────────────────────────────
router.patch('/:id/unread', authenticate, async (req, res) => {
  try {
    const { rowCount } = await pool.query(
      `UPDATE notifications SET read = FALSE WHERE id = $1 AND user_id = $2`,
      [req.params.id, req.user.id]
    );
    if (rowCount === 0) return res.status(404).json({ error: 'Not found' });
    res.json({ status: 'unread' });
  } catch (err) {
    console.error('[notifications] PATCH unread error:', err.message);
    res.status(500).json({ error: 'Server error' });
  }
});

// ── PATCH /api/notifications/read-all/mark ───────────────────────────────────
router.patch('/read-all/mark', authenticate, async (req, res) => {
  try {
    const { rowCount } = await pool.query(
      `UPDATE notifications SET read = TRUE WHERE user_id = $1 AND read = FALSE`,
      [req.user.id]
    );
    res.json({ updated: rowCount });
  } catch (err) {
    console.error('[notifications] PATCH read-all error:', err.message);
    res.status(500).json({ error: 'Server error' });
  }
});

// ── PATCH /api/notifications/bulk ────────────────────────────────────────────
// Body: { ids: string[], read: boolean }
router.patch('/bulk', authenticate, async (req, res) => {
  const { ids, read } = req.body;
  if (!Array.isArray(ids) || ids.length === 0) return res.status(400).json({ error: 'ids required' });
  if (typeof read !== 'boolean') return res.status(400).json({ error: 'read must be boolean' });
  try {
    const { rowCount } = await pool.query(
      `UPDATE notifications SET read = $1 WHERE id = ANY($2::uuid[]) AND user_id = $3`,
      [read, ids, req.user.id]
    );
    res.json({ updated: rowCount });
  } catch (err) {
    console.error('[notifications] PATCH bulk error:', err.message);
    res.status(500).json({ error: 'Server error' });
  }
});

// ── POST /api/notifications/sla-breach ───────────────────────────────────────
router.post('/sla-breach', authenticate, async (req, res) => {
  const { ticket_id, customer_name, assigned_to, priority = 3 } = req.body;
  if (!ticket_id) return res.status(400).json({ error: 'ticket_id required' });

  const tierLabel = { 1: 'VIP', 2: 'EA', 3: 'Standard' }[priority] ?? 'Standard';
  const customerLabel = customer_name || 'Unknown customer';
  const title = `SLA Breached — Ticket #${ticket_id.slice(0, 8)}`;
  const body  = `${customerLabel} (${tierLabel}) — response time exceeded`;

  const created = [];
  try {
    // Notify assigned agent
    if (assigned_to) {
      const id = uuidv4();
      const { rows } = await pool.query(
        `INSERT INTO notifications (id, user_id, role, type, priority, title, body, ticket_id)
         VALUES ($1,$2,'agent','sla_breach','critical',$3,$4,$5)
         RETURNING id, user_id, role, type, priority, title, body, ticket_id, read, created_at`,
        [id, assigned_to, title, body, ticket_id]
      );
      created.push(rows[0]);
      emitToAgent(assigned_to, 'notification:new', { notification: rows[0] });
    }

    // Notify all supervisors
    const { rows: supers } = await pool.query(
      `SELECT id FROM users WHERE role IN ('supervisor','admin','super_admin') AND active = TRUE`
    );
    for (const s of supers) {
      const id = uuidv4();
      const { rows } = await pool.query(
        `INSERT INTO notifications (id, user_id, role, type, priority, title, body, ticket_id)
         VALUES ($1,$2,'supervisor','sla_breach','critical',$3,$4,$5)
         RETURNING id, user_id, role, type, priority, title, body, ticket_id, read, created_at`,
        [id, s.id, title, body, ticket_id]
      );
      created.push(rows[0]);
      emitToAgent(s.id, 'notification:new', { notification: rows[0] });
    }

    res.json({ created: created.length });
  } catch (err) {
    console.error('[notifications] sla-breach error:', err.message);
    res.status(500).json({ error: 'Server error' });
  }
});

module.exports = router;
